core/platform/auth/firebase_auth_services.py

```diff
@@ -0,0 +1,94 @@
+# coding: utf-8
+#
+# Copyright 2020 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Service layer for handling the authentication details of users via Firebase.
+
+Oppia depends upon the OpenID Connect 1.0 specification for handling user
+authentication, and uses [Firebase](https://firebase.google.com/docs/auth) as
+the primary provider.
+
+Glossary:
+    OpenID Connect 1.0: a specification (i.e., a strict set of algorithms, data
+        structures, and rules) that defines how two parties must share data
+        about a user securely, on that user's behalf.
+    Claim: a piece of information about a user (name, address, phone number,
+        etc.) that has been encrypted and digitally signed.
+    JWT: JSON Web Token; a compact and URL-safe protocol primarily designed to
+        send Claims between two parties. Claims are organized in a JSON object
+        that maps "Claim Name" to "Claim Value".
+    Identity provider: An entity that creates, maintains, and manages identity
+        information and provides authentication services. Such services rely on
+        JWTs to send identity information. Examples of identity providers
+        include: Google, Facebook, Email verification links, and Text message
+        SMS codes.
+"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import contextlib
+
+from core.platform import models
+
+import firebase_admin
+from firebase_admin import auth
+
+(auth_models,) = models.Registry.import_models([models.NAMES.auth])
+
+UserIdBySubjectIdModel = auth_models.UserIdByFirebaseSubjectIdModel  # pylint: disable=invalid-name
```

Comment:
I don't think you need to (or should) alias this. Also, you don't use it in this file?

Reply:
It feels like this a Firebase-specific implementation detail, so that's why I decided to put it here.

Put in a different view, clients of the `get_user_id_by_subject_id` shouldn't care that the model is specific to Firebase. That's why I felt like it should be kept alongside the other firebase-specific services (which is the point of this file, right?)

WDYT?

Reply:
I think a different implementation would use a different linkage model, so I think it's fine to keep Firebase in the model name.

Reply:
I still don't like the idea of exposing `get_user_id_by_firebase_subject_id` or `UserIdByFirebaseSubjectId`, it feels like a leakage / like the abstraction is wrong.

I could move `get_user_id_by_subject_id` into this file instead, but then the "you must acquire a context" restriction might become too overbearing.

Maybe I'm using the wrong abstraction here..?

Reply:
Yup, I get the concern. How about get_user_id_by_auth_id and UserIdByAuthIdModel? I think "auth" is the general word that you might be looking for.

Reply:
Done, I've gone with a function named `authenticate_request`.